### ISSUE TITLE
C++: switch types leak memory

### ISSUE THREAD

**Issue by GreyCat** [kaitai-io/kaitai_struct#209]
It turns out that current compiler generates the following code, for example, for [switch_manual_int](https://github.com/kaitai-io/kaitai_struct_tests/blob/master/formats/switch_manual_int.ksy):

```cpp
switch_manual_int_t::opcode_t::opcode_t(kaitai::kstream *p_io, switch_manual_int_t* p_parent, switch_manual_int_t *p_root) : kaitai::kstruct(p_io) {
    m__parent = p_parent;
    m__root = p_root;
    _read();
}

void switch_manual_int_t::opcode_t::_read() {
    m_code = m__io->read_u1();
    switch (code()) {
    case 73:
        m_body = new intval_t(m__io, this, m__root);
        break;
    case 83:
        m_body = new strval_t(m__io, this, m__root);
        break;
    }
}

switch_manual_int_t::opcode_t::~opcode_t() {
}
```

That is, `m_body` is:

* Not initialized properly (there should be `m_body = 0` initializer in the constructor)
* Not deleted properly (there should be some conditional `delete` in the destructor)

What's the worst, `valgrind-cpp_stl` seems to be totally silent about all this.

**Comment by GreyCat** [comment]
Cc @scottmsilver, who reported this in the first place.

**Comment by GreyCat** [comment]
Actually, I was wrong, valgrind detects this (and a whole lot of memory leaks) properly.

**Comment by GreyCat** [comment]
Ok, the baseline (what we have right now) is:

```
==26488== LEAK SUMMARY:
==26488==    definitely lost: 4,534 bytes in 54 blocks
==26488==    indirectly lost: 4,048 bytes in 22 blocks
==26488==      possibly lost: 0 bytes in 0 blocks
==26488==    still reachable: 0 bytes in 0 blocks
==26488==         suppressed: 0 bytes in 0 blocks
==26488== 
==26488== For counts of detected and suppressed errors, rerun with: -v
==26488== ERROR SUMMARY: 42 errors from 38 contexts (suppressed: 0 from 0)
```

**Comment by scottmsilver** [comment]
Not sure what you're thinking but probably the simplest way to solve this is to wrap m__io__raw_body and m_body in unique_ptr's. It cleans up the destructor, avoids that constructor initialization to NULL you were thinking of, too.  In general, given that the choice seems to be be eager reading, feels like you could go a step further and use the principle of RAII if you'd like.

**Comment by JaapAap** [comment]
I was having similar thoughts. What C++ version is targeted? There are some comments in the code on "this works for C++11 only and needs to be fixed" but using smart pointers really simplifies the logic in some cases.

**Comment by GreyCat** [comment]
Generally, the original idea was to:

1. Target very early C++ versions, like C++98 by default, to be universally working everywhere, even in pretty dated environments
2. Add support for more modern C++ versions later (like C++11 or C++17), using all the modern stuff like `std::move`, `unique_ptr`, `shared_ptr`, etc

I guess now the most straightforward answer to this is to fix destructor creation. Introducing smart pointers is a huge leap ahead, but it will require throughtout thinking about and careful planning.

**Comment by JaapAap** [comment]
Could such a thing be done gradually, e.g. by introducing a separate C++11 compiler option and using the current CPP implementation by default while overriding some of its methods?

**Comment by GreyCat** [comment]
As far as I understand, it's a pretty big change anyway. You'll need to change:

* all the data types (i.e. from `foo_t*` to something like `std::shared_ptr<foo_t>`,
* all the instancing (i.e. all the stuff would be probably now happen on the stack instead of on the heap, no more `new` / `delete` calls, but sometimes one would need more intricate stuff like `std::make_shared`, `std::move`, etc)
* all the usage (i.e. no more `foo()->bar()->baz()`, probably it would switch to something more specific)

**Comment by JaapAap** [comment]
I haven't thought this through, but what would you mean by 'more specific' usage?

**Comment by GreyCat** [comment]
Well, I'm not like a C++ expert here, I'm always consulting those who really use it in real-life, so it's not like a have a full picture in my head as well.

I see that there's some interest in making a C++11/C++17 target for a compiler, hang on a sec, I'll create a separate issue for that, so all these ideas would be kept in one place (instead of being in an issue dedicated to memory leaks).

**Comment by GreyCat** [comment]
Here we go, #211.

**Comment by GreyCat** [comment]
First batch of updates that should close many, but not all C++ memory leaks is pushed and is gone [to CI](http://kaitai.io/ci/). Results should be there in several minutes.

* Previous results: 83% tests pass, 18 tests show memory leaks
* Expected results now: 92% pass, 7 memory leaks remain

**Comment by GreyCat** [comment]
Next major problem: generated IO stream object needs to be stored and taken care of properly as well. For example, [this new](https://github.com/kaitai-io/ci_targets/blob/master/compiled/cpp_stl/instance_user_array.cpp#L59) has no corresponding delete in [the destructor](https://github.com/kaitai-io/ci_targets/blob/master/compiled/cpp_stl/instance_user_array.cpp#L20-L28).

**Comment by GreyCat** [comment]
Ok, this seems to be solved, off to CI it goes.

Expected results now: 94% pass, 5 leaks remain.

**Comment by GreyCat** [comment]
Well, looks like I've fixed the most obvious problems, but Valgrind still reports leaks. I'm kind of puzzled at what's wrong with SwitchManualIntSize right now, for example. Anyone can take a look?

**Comment by JaapAap** [comment]
Could it be that these are false positives (see http://valgrind.org/docs/manual/faq.html#faq.reports point 4.1)? What is reported exactly?

**Comment by GreyCat** [comment]
@JaapAap I doubt that they're false positives.

Here's a [section of Valgrind report](https://github.com/kaitai-io/ci_artifacts/blob/cpp_stl/gcc4.8_linux/test_out/cpp_stl/valgrind.xml#L147-L242) that mentions leak near [switch_manual_int_size.cpp:71](https://github.com/kaitai-io/ci_targets/blob/master/compiled/cpp_stl/switch_manual_int_size.cpp#L71).

Probably the culprit is actually somewhere either in [~chunk_t](https://github.com/kaitai-io/ci_targets/blob/master/compiled/cpp_stl/switch_manual_int_size.cpp#L56-L61) or [~switch_manual_int_size_t](https://github.com/kaitai-io/ci_targets/blob/master/compiled/cpp_stl/switch_manual_int_size.cpp#L20-L25) or [~chunk_dir_t](https://github.com/kaitai-io/ci_targets/blob/master/compiled/cpp_stl/switch_manual_int_size.cpp#L90-L92) — but all 3 seem to look legit for my untrained eye...

**Comment by scottmsilver** [comment]
The valgrind report seems to be the FAQ item that JaapAap send a link to.
Can you try adding that flag to gcc to see if it goes away?

Scott

**Comment by scottmsilver** [comment]
I believe you have to make the destructor ~kstruct() virtual. Not sure what
Valgrind is pointing at. Using MacOS Instruments yields no leaks after this
for my tests.

**Comment by GreyCat** [comment]
I've tried the flags, and it doesn't seem to change anything in the output. I believe that these flags are meant to avoid some internal glibc/libstdc++ false positives, and here it precisely pinpoints some places in our generated code, not in some obscure system libraries locations.

Indeed, a suggestion to make a virtual destructor seems to be the bull's eye :) I've just learned that without it, it's considered undefined behavior, and both GCC and Clang actually just silently ignore destructor of derived class. I'll submit this fix now and we'll see how it will turn out at CI. A ran at my home box now shows just one leak in Valgrind — in default_endian_expr_exception test.
